llvm/lib/Transforms/InstCombine/InstCombineCalls.cpp

```diff
@@ -3761,6 +3762,24 @@ Instruction *InstCombinerImpl::visitCallInst(CallInst &CI) {
             return replaceInstUsesWith(CI, Res);
           }
       }
+
+      // Handle the case where a splat is summarized. In that case we have a
+      // multpilication. For example: %2 = insertelement <4 x i32> poison, i32
+      // %0, i64 0 %3 = shufflevector <4 x i32> %2, poison, <4 x i32>
+      // zeroinitializer %4 = tail call i32 @llvm.vector.reduce.add.v4i32(%3)
+      // =>
+      // %2 = mul i32 %0, 4
+      if (Value *Splat = getSplatValue(Arg)) {
+        // It is only a multiplication if we add the same element over and over.
+        ElementCount ReducedVectorElementCount =
+            cast<VectorType>(Arg->getType())->getElementCount();
+        if (ReducedVectorElementCount.isFixed()) {
```

Comment:
Can you please add a negative test with a scalable vector? (We could handle it as well, but it's a bit trickier.)

Reply:
CreateElementCount() doesn't have the right overflow behavior. It generates a mul nuw, and llvm.vscale is poison on overflow. I guess we could do something like generate an i64 CreateElementCount and then zext or trunc to the final type.